Graph-iOS-Swift-Snippets/AuthenticationProvider.swift

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license.
+ * See LICENSE in the project root for license information.
+ */
+
+import Foundation
+import MSAL
+import MSGraphSDK
+
+class AuthenticationProvider: NSObject, MSAuthenticationProvider
+{
+    var accessToken: String?
+    var account: MSALAccount?
+    var msalClient: MSALPublicClientApplication
+    
+    init(clientId: String, authority: MSALAuthority) throws
+    {
+        let config = MSALPublicClientApplicationConfig(clientId: clientId)
+        msalClient = try MSALPublicClientApplication(configuration: config)
+    }
+    
+    func acquireAuthToken(scopes: [String], completion:@escaping (_ success:Bool, _ error: NSError?) -> Void)
+    {
+        var completionBlock: MSALCompletionBlock?
+        completionBlock = { (result, error) in
+            
+            if let error = error as NSError?
+            {
+                if (error.domain == MSALErrorDomain)
+                {
+                    let msalError = MSALError(rawValue: error.code)!
+                    
+                    switch msalError {
+                    case MSALError.interactionRequired:
+                        let interactiveParameters = MSALInteractiveTokenParameters(scopes: scopes)
+                        self.msalClient.acquireToken(with: interactiveParameters, completionBlock: completionBlock!)
+                        
+                    case MSALError.serverDeclinedScopes:
+                        var declinedScopes = [String]()
+                        if let ds = error.userInfo[MSALDeclinedScopesKey] as? Array<String> { declinedScopes = ds }
+                        print("The following scopes were declined: ", declinedScopes)
+                        
+                        var grantedScopes = [String]()
+                        if let gs = error.userInfo[MSALGrantedScopesKey] as? Array<String> { grantedScopes = gs }
+                        print("Trying to acquire a token with granted scopes: ", grantedScopes)
+                        
+                        var accounts = [MSALAccount]()
+                        do {
+                            accounts = try self.msalClient.allAccounts()
+                        } catch let error as NSError {
+                            print("Error: ", error)
+                            completion(false, error)
+                            return
+                        }
+                        
+                        if let account = accounts.first {
```

Comment:
should we have a separate function to call acquire token silent with parameters? Would make this error handling a bit more readable.

Reply:
done